File: onyx-database/src/main/java/com/onyx/diskmap/node/SkipListHeadNode.java
package com.onyx.diskmap.node;

import com.onyx.buffer.BufferStream;
import com.onyx.buffer.BufferStreamable;
import com.onyx.exception.BufferingException;
import com.onyx.persistence.context.SchemaContext;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Created by tosborn1 on 1/9/17.
 *
 * This is a head of a skip list level.
 */
public class SkipListHeadNode implements BufferStreamable {

    public static final int HEAD_SKIP_LIST_NODE_SIZE = Long.BYTES * 2 + Byte.BYTES;

    public byte level;
    public long next;
    public long down;
    public long position;

    public SkipListHeadNode()
    {

    }

    public SkipListHeadNode(byte level, long next, long down)
    {
        this.level = level;
        this.next = next;
        this.down = down;
    }

    @Override
    public void read(BufferStream buffer) throws BufferingException {
        next = buffer.getLong();
        down = buffer.getLong();
        level = buffer.getByte();
//        position = buffer.getLong();
    }

    @Override
    public void write(BufferStream buffer) throws BufferingException {
        buffer.putLong(next);
        buffer.putLong(down);
        buffer.putByte(level);
//        buffer.putLong(position);
    }

    @Override
    public void write(@NotNull BufferStream buffer, @Nullable SchemaContext context) throws BufferingException {
        write(buffer);
    }

    @Override
    public void read(@NotNull BufferStream buffer, @Nullable SchemaContext context) throws BufferingException {
        read(buffer);
    }

    @Override
    public boolean equals(Object o) {
        return (o instanceof SkipListHeadNode && ((SkipListHeadNode) o).position == position);
    }

    @Override
    public int hashCode() {
        return (int)(position ^ (position >>> 32));
    }
}

File: onyx-database/src/main/kotlin/com/onyx/diskmap/data/SkipListHeadNode.kt
package com.onyx.diskmap.data

import com.onyx.buffer.BufferStream
import com.onyx.buffer.BufferStreamable
import com.onyx.exception.BufferingException

/**
 * Created by tosborn1 on 1/9/17.
 *
 * This is a head of a skip list level.
 */
open class SkipListHeadNode : BufferStreamable {

    var level: Byte = 0
    var next: Long = 0
    var down: Long = 0
    var position: Long = 0

    constructor()

    constructor(level: Byte, next: Long, down: Long) {
        this.level = level
        this.next = next
        this.down = down
    }

    @Throws(BufferingException::class)
    override fun read(buffer: BufferStream) {
        next = buffer.long
        down = buffer.long
        level = buffer.byte
    }

    @Throws(BufferingException::class)
    override fun write(buffer: BufferStream) {
        buffer.putLong(next)
        buffer.putLong(down)
        buffer.putByte(level)
    }

    override fun equals(other: Any?): Boolean = other is SkipListHeadNode && other.position == position
    override fun hashCode(): Int  = position.hashCode()

    companion object {

        val HEAD_SKIP_LIST_NODE_SIZE = java.lang.Long.BYTES * 2 + java.lang.Byte.BYTES
    }
}
